pkg/cri/server/sandbox_run.go

```diff
@@ -95,6 +84,40 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		},
 	)
 
+	// cleanupErr records the last error created by the critical cleanup in defer functions,
+	// like CNI teardown and stop running sandbox task.
+	// So when the cleanup is not completed for some reason, the CRI-plugin will remain the sandbox
+	// in the not-ready state, which can be cleanup by kubelet syncPod.
+	var cleanupErr error
+
+	defer func() {
+		if cleanupErr != nil {
+			log.G(ctx).WithError(cleanupErr).Errorf("there was error during resource cleanup and the sandbox will be put into store")
+
+			sandbox.Status = sandboxstore.StoreStatus(
+				sandboxstore.Status{
+					State: sandboxstore.StateNotReady,
+				},
+			)
+			if err := c.sandboxStore.Add(sandbox); err != nil {
+				log.G(ctx).WithError(err).Errorf("failed to add sandbox %+v into store", sandbox)
+			}
+		}
+	}()
+
+	// Reserve the sandbox name to avoid concurrent `RunPodSandbox` request starting the
+	// same sandbox.
+	if err := c.sandboxNameIndex.Reserve(name, id); err != nil {
```

Comment:
Why moving this down?

Reply:
you mean move this defer function below c.sandboxNameIndex.Reserve?

I think both way have equivalent logic. For readability, I make the this "persist sandbox/container if cleanup fails" at the top and put sandboxNameIndex.Reserve and sandboxNameIndex.ReleaseByName together

Reply:
ok, so you prefer:

```
err = sandboxNameIndex.Reserve
if err != nil {
  return
}

defer func() {
  // put sandbox into store
}()

defer func() {
   if retErr != nil && cleanupErr == nil {
		c.sandboxNameIndex.ReleaseByName(name)
   }
}
```

Reply:
Talked offline, I will move the `sandboxNameIndex.Reserve` upwards

Reply:
No, the plan is to keep the order in current way.

Reply:
Sorry for confusing you. Actually I did move down the "putting sandbox into store"